src/EasyNetQ.HostedService/RabbitMqProducer.cs

```diff
@@ -249,55 +283,65 @@ private protected virtual void OnCancellation()
+                            $"{RabbitMqConfig.Id} ({GetMessageInformation(message)})";
 
-                        ProducerLoopWaitAndContinue(cancellationToken);
+                        FailAndDiscardMessage(message, new Exception(error, exception));
                     }
                     catch (AlreadyClosedException exception)
                     {
-                        Logger?.LogError(
+                        var error =
                             $"AMQP error in producer loop: {exception.Message}\n{exception.StackTrace}\n" +
-                            $"({GetMessageInformation(message)})");
+                            $"({GetMessageInformation(message)})";
 
-                        // 404 - NOT FOUND means that the message cannot be processed at all at this point and it's
-                        // best to notify the library's client and discard it
-                        if (exception.ShutdownReason?.ReplyCode == 404)
-                        {
-                            ProducerLoopDiscardMessageAndContinue(message);
-                        }
+                        FailAndDiscardMessage(message, new Exception(error, exception));
                     }
                     catch (Exception exception)
                     {
-                        Logger?.LogCritical(
+                        var error =
                             $"Critical error in producer loop: {exception.Message}\n{exception.StackTrace}\n" +
-                            $"({GetMessageInformation(message)})");
+                            $"({GetMessageInformation(message)})";
 
-                        ProducerLoopWaitAndContinue(cancellationToken);
+                        FailAndDiscardMessage(message, new Exception(error, exception));
+                    }
+                    finally
+                    {
+                        if (success)
+                        {
+                            message.TracingActivity?.Dispose();
+                        }
                     }
                 }
             }).Start();
 
-        private void ProducerLoopWaitAndContinue(CancellationToken cancellationToken)
+        private void FailAndDiscardMessage(Message message, Exception exception)
         {
-            Task.Delay(RabbitMqConfig.PublisherLoopErrorBackOffMilliseconds, cancellationToken).Wait(cancellationToken);
+            if (message?.TracingActivity != null)
+            {
+                message.TracingActivity.AddEvent(new ActivityEvent(TraceEventName.Exception,
+                    tags: new ActivityTagsCollection(new[]
+                        { new KeyValuePair<string, object>("exception", exception) })));
 
-            _messageSemaphore.Release();
-        }
+                message.TracingActivity.Dispose();
+            }
 
-        private void ProducerLoopDiscardMessageAndContinue(Message message)
-        {
+            message?.TaskCompletionSource.SetException(exception);
```

Comment:
If you go through the failanddiscard message option above you will need to change this to TrySetException